### ISSUE TITLE
TOTP multifactor authentication breaks Oauth logins

### ISSUE THREAD

**Issue by Gummikavalier** [RocketChat/Rocket.Chat#29643]
### Description:

Oauth / OIDC logins fail when time based one-time password (TOTP) is enabled for the user.

### Steps to reproduce:

1. Set up custom Oauth config for logins
2. As a regular user enable and configure TOTP into use
3. Log out
4. Log back in
5. Fill up the TOTP code

### Expected behavior:

User is queried for the TOTP code.

### Actual behavior:

The user never gets to a point of filling up the TOTP code. Instead they get thrown back to the login screen.

### Server Setup Information:

- Version of Rocket.Chat Server: 6.2.8 Enterprise
- Operating System: RHEL8
- Deployment Method: tar
- Number of Running Instances: 10
- DB Replicaset Oplog: Yes
- NodeJS Version: 14.21.3
- MongoDB Version: 5.0

### Client Setup Information

- Desktop App or Browser Version: All browsers

### Additional context

Errors:

![Screenshot from 2023-06-26 10-05-07](https://github.com/RocketChat/Rocket.Chat/assets/6958942/1e8318eb-cf3f-4a46-a765-b8e31f0d1414)

Browser log shows one uncaught exception:
```
adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:5 fuselage: 0.31.23
adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:1599 MessageType.render is deprecated. Use MessageType.message instead. livechat_webrtc_video_call
registerType @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:1599
e @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:1599
g @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:3
r @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:3
a @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:5
e @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:1599
g @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:3
r @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:3
a @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:5
e @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:1599
g @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:3
r @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:3
a @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:5
e @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:1599
g @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:3
r @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:3
t @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:3
(anonymous) @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:1599
adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:1167 strict mode: missing type "object" for keyword "additionalProperties" at "#" (strictTypes)
k @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:1167
q @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:1164
z @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:1164
R @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:1164
M @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:1164
C @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:1164
(anonymous) @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:1164
code @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:1167
(anonymous) @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:1164
code @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:1167
func @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:1167
h @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:1164
_ @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:1164
g @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:1164
u @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:1167
_compileSchemaEnv @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:1164
compile @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:1164
e @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:1102
g @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:3
r @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:3
a @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:5
e @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:1102
g @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:3
r @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:3
a @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:5
e @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:1102
g @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:3
r @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:3
a @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:5
e @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:1599
g @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:3
r @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:3
a @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:5
e @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:1599
g @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:3
r @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:3
a @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:5
e @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:1599
g @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:3
r @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:3
t @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:3
(anonymous) @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:1599
adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:1167 strict mode: missing type "object" for keyword "properties" at "#" (strictTypes)
k @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:1167
q @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:1164
z @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:1164
R @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:1164
M @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:1164
C @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:1164
(anonymous) @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:1164
code @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:1167
(anonymous) @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:1164
code @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:1167
func @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:1167
h @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:1164
_ @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:1164
g @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:1164
u @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:1167
_compileSchemaEnv @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:1164
compile @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:1164
e @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:1102
g @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:3
r @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:3
a @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:5
e @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:1102
g @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:3
r @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:3
a @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:5
e @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:1102
g @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:3
r @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:3
a @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:5
e @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:1599
g @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:3
r @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:3
a @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:5
e @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:1599
g @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:3
r @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:3
a @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:5
e @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:1599
g @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:3
r @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:3
t @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:3
(anonymous) @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:1599
adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:1167 strict mode: use allowUnionTypes to allow union type keyword at "#/properties/value" (strictTypes)
k @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:1167
q @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:1164
N @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:1164
R @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:1164
M @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:1164
C @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:1164
E @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:1164
A @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:1164
subschema @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:1164
m @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:1170
code @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:1170
V @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:1164
(anonymous) @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:1164
code @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:1167
block @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:1167
F @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:1164
h @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:1164
(anonymous) @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:1164
code @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:1167
block @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:1167
M @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:1164
C @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:1164
(anonymous) @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:1164
code @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:1167
(anonymous) @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:1164
code @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:1167
func @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:1167
h @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:1164
_ @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:1164
g @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:1164
u @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:1167
_compileSchemaEnv @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:1164
compile @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:1164
e @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:1102
g @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:3
r @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:3
a @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:5
e @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:1102
g @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:3
r @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:3
a @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:5
e @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:1599
g @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:3
r @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:3
a @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:5
e @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:1599
g @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:3
r @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:3
a @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:5
e @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:1599
g @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:3
r @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:3
t @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:3
(anonymous) @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:1599
adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:1500 Uncaught (in promise) undefined
onStop @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:1500
_livedata_nosub @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:1410
onMessage @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:1410
(anonymous) @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:1380
forEachCallback @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:1380
socket.onmessage @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:1380
r.dispatchEvent @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:1380
k._dispatchMessage @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:1380
k._didMessage @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:1380
o.ws.onmessage @ adbe793e41282308cffbfe2690d36f051a7ffdf4.js?meteor_js_resource=true:1380
```

**Comment by Gummikavalier** [comment]
For those who want to workaround the issue, you can disable TOTP per user by running:
```
db.users.update( {'username': 'insertusernamehere'}, {$set: {'services.totp': 'false'}} )
```
in the database.

**Comment by Gummikavalier** [comment]
Issue still active in RC 6.3.0.

**Comment by casalsgh** [comment]
Internal squads checking into this one

**Comment by matheusbsilva137** [comment]
What OAuth provider are you using? @Gummikavalier 
I guess this is not happening to all of them :thinking:

**Comment by pierre-lehnen-rc** [comment]
Anybody had this issue happen with the OAuth login style configured to "Popup"?

So far I've only managed to observe this bug when that setting is set to "Redirect".

**Comment by Gummikavalier** [comment]
@pierre-lehnen-rc Thanks for the fix. :heart:

I did tests and it works on Firefox (clean session too). On Chrome, at least on clean sessions, it still exists.

Usually we need to use the redirect method to comply with the security policies, as any popups always open up a big logical spoofing vector into the login process. But we'll switch to popup for time being if RC 6.4.0 will come with another fix.

We are using a custom Oauth provider.

**Comment by Gummikavalier** [comment]
Looks like RC 6.3.2 fixed the redirected page issue with Chrome _or_ it was just something else earlier. Seems to work now.

**Comment by Gummikavalier** [comment]
I did some testing and with redirect method TOTP window works often after the second login try using the original Chrome incognito browser session. But it is not guaranteed.

Also the failed login banning system masks these errors somewhat; after too many bad login attempts the outcome of the banned logins look exactly the same (loop back to the login screen).

I had to disable failed login IP banning in the settings to confirm that the issue still remains with the redirect method.
